### DISCUSSION TITLE
Copybooks in repository are not moving to PDS, Compilation is using syslib copybooks in cobol.properties concatenation

### DISCUSSION THREAD

**Discussion by APSDevp** [IBM/dbb-zappbuild#384] [General]
COBOL copybooks are taken from syslib concatenation in cobol.properties and not from COPYBOOK folder in the repository. Copybooks are moved to USS but not copied to PDS in our case we modified to User.copybook.

**Comment by dennis-behm** [comment]
@APSDevp, this does not look like a zAppBuild issue.

Can you please check, if the `SearchPathDependencyResolver` is resolving the local files? Also check your SearchPath configuration which is used by DBB to locate the copybooks.

In the verbose output, you should find something like this:

```
*** Resolution rules for MortgageApplication/cobol/epsnbrvl.cbl:
search:/u/dbehm/userBuild/?path=MortgageApplication/copybook/*.cpy
*** Physical dependencies for MortgageApplication/cobol/epsnbrvl.cbl:
{
  "lname": "EPSNBRPM",
  "library": "SYSLIB",
  "category": "COPY",
  "sourceDir": "/u/dbehm/userBuild/",
  "file": "MortgageApplication/copybook/epsnbrpm.cpy",
  "resolved": true,
  "excluded": false
}
```

**Comment by dennis-behm** [comment]
Converting this into a discussion

**Comment by APSDevp** [comment]
@dennis-behm 

Thanks for your quick response! I have tried checking with verbose mode and my findings are

     1) Script is not using the SearchPathDependencyResolver it is using the deprecated resolver with the search path of sourceDir and Directory where the copybooks are present and it has used the default scanner, however, couldn't able to see the "file" : parameter where copybooks are stored as example you have shared 
 
 *** Physical dependencies for {app'name_/cobol/{prog_name}.cbl
{excluded:false,lnanme:"{cpybookName},"library": "SYSLIB", "category": "COPY",**"resolved": false**,}  <-- likewise am getting.

     2) In cobol.groovy at present in this repository below piece of code is there

           // configure dependency resolution and create logical file	
	String dependencySearch = props.getFileProperty('cobol_dependencySearch', buildFile)
	SearchPathDependencyResolver dependencyResolver = new SearchPathDependencyResolver(dependencySearch)

       cobol.groovy which am using has different set of codes which uses the resolverUtils and  buildutils( deprecated one which my scripts invokes using)

     Could you please help me on this to resolve it?

**Reply by APSDevp** [reply]
@dennis-behm 
Tried with SearchPathDependencyResolver by changing the property and build toolkit version and in cobol.groovy now it's taking the SearchPathDependencyResolver but still the dependency resolution is showing false "resolved: false", could you please help me on what I am missing here

**Reply by dennis-behm** [reply]
@APSDevp,  let's take a step back. The `file` attribute is only available in the output, when it is resolved! Otherwise it is an unresolved dependency.

There are two different APIs - the SearchPathDependencyResolver, which is available since DBB 1.1.2. Starting with DBB 2.0, this is the only available API.

The DependencyResolver is only available on DBB 1.x, and is marked as deprecated, and is replaced by SearchPathDependencyResolver.

The old DependencyResolver API, requires to configure the pseudo JSON - 
https://github.com/IBM/dbb-zappbuild/blob/54abed229b647970eda2ad7a0cd2650065e17ff1/samples/application-conf/application.properties#L150-L158

The new SearchPathDependencyResolver API is using the new schema for the search path - 
https://github.com/IBM/dbb-zappbuild/blob/54abed229b647970eda2ad7a0cd2650065e17ff1/samples/application-conf/application.properties#L160-L182

Is your searchPathConfiguration correctly set up? And do you see the second line in the verbose output?
```
*** Resolution rules for MortgageApplication/cobol/epsnbrvl.cbl:
search:/u/dbehm/userBuild/?path=MortgageApplication/copybook/*.cpy
```
Can you share the section of your verbose output? If you cannot share it here, please open an official support ticket with IBM.

**Reply by APSDevp** [reply]
hi @dennis-behm 

Yes able to see similar to this 
*** Resolution rules for MortgageApplication/cobol/epsnbrvl.cbl:
search:/u/dbehm/userBuild/?path=MortgageApplication/copybook/*.cpy     --> under this path there was a copybook folder with copybooks but still not copied to PDS

Next line is *** physical dependencies {application/COBOL/{progname}

{excluded:false,lnanme:"{cpybookName},"library": "SYSLIB", "category": "COPY","resolved": false,} 

whether am missing any Properties?

**Reply by dennis-behm** [reply]
Hi, 

With the provided information I can only guess which makes it difficult to help - without any log, or seeing your concrete configuration and checking your system. I would recommend that you open an IBM support ticket.

One last guess is that the filter is not formed correctly for the dependency resolution.

Please run an `ls` like below with your filter on USS:
```
DBEHM:/u/dbehm/userBuild: >ls -lisa MortgageApplication/**/*.cpy
17247   16 -rw-------   1 BPXROOT  TIVUSR       577 Jan 10  2023 MortgageApplication/copybook/epsmortf.cpy
17250   16 -rw-------   1 BPXROOT  TIVUSR       104 Mar 30 13:51 MortgageApplication/copybook/epsmtcom.cpy
17251   16 -rw-------   1 BPXROOT  TIVUSR       333 Mar 30 13:51 MortgageApplication/copybook/epsmtinp.cpy
 3639   16 -rw-------   1 BPXROOT  TIVUSR       399 Mar 30 13:51 MortgageApplication/copybook/epsmtout.cpy
17252   16 -rw-------   1 BPXROOT  TIVUSR       386 Mar 30 13:51 MortgageApplication/copybook/epsnbrpm.cpy
17253   16 -rw-------   1 BPXROOT  TIVUSR       534 Mar 30 13:51 MortgageApplication/copybook/epspdata.cpy
15535    2 -rw-------   1 BPXROOT  TIVUSR        32 May 10 15:27 MortgageApplication/copybook/samp.cpy
```
Do you use any `**` patterns in the filter?

**Reply by APSDevp** [reply]
@dennis-behm 

When I tried to like this "ls -lisa MortgageApplication/**/*.cpy" - able to see all the copybooks.

But when I tried the same with  "ls -lisa MortgageApplication/copybook/*.cpy " it showed an invalid directory.

As you have guessed - the filter is not formed correctly.

Do I need to change -> copybookSearch = search:${workspace}/?path=${application}/copybook/*.cpy to copybookSearch = search:${workspace}/?path=${application}/**/*.cpy  will it work ?

**Reply by dennis-behm** [reply]
@APSDevp , you need provide  the correct path that your application is using which can be `copybook` but also any other name.

Don't know if this is a typo in the msg `?path=${application}/copybook/.cpy` but it misses the `*` before `.cpy`

Also `MortgageApplication/copybook/*.cpy` works fine:

```
DBEHM:/u/dbehm/userBuild: >ls -lisa MortgageApplication/copybook/*.cpy
17247   16 -rw-------   1 BPXROOT  TIVUSR       577 Jan 10  2023 MortgageApplication/copybook/epsmortf.cpy
17250   16 -rw-------   1 BPXROOT  TIVUSR       104 Mar 30 13:51 MortgageApplication/copybook/epsmtcom.cpy
17251   16 -rw-------   1 BPXROOT  TIVUSR       333 Mar 30 13:51 MortgageApplication/copybook/epsmtinp.cpy
 3639   16 -rw-------   1 BPXROOT  TIVUSR       399 Mar 30 13:51 MortgageApplication/copybook/epsmtout.cpy
17252   16 -rw-------   1 BPXROOT  TIVUSR       386 Mar 30 13:51 MortgageApplication/copybook/epsnbrpm.cpy
17253   16 -rw-------   1 BPXROOT  TIVUSR       534 Mar 30 13:51 MortgageApplication/copybook/epspdata.cpy
15535    2 -rw-------   1 BPXROOT  TIVUSR        32 May 10 15:27 MortgageApplication/copybook/samp.cpy
```

**Comment by APSDevp** [comment]
@dennis-behm 

The issue here is the directory structure is "COPYBOOK" but the script searching for "copybook", after changing it - now able to see the repository copybook in the PDS, thanks for your support - Have learned a lot in this journey.
